WIP: feat: usename type in access

Aflynn50 opened the pull request:
Switch to using a type from core/user for usernames to fix issues with external users

<!-- 
The PR title should match: <type>(optional <scope>): <description>.

Please also ensure all commits in this PR comply with our conventional commits specification:
https://docs.google.com/document/d/1SYUo9G7qZ_jdoVXpUVamS5VCgHmtZ0QA-wZxKoMS-C0 
-->

<!-- Why this change is needed and what it does. -->

## Checklist

<!-- If an item is not applicable, use `~strikethrough~`. -->

- [ ] Code style: imports ordered, good names, simple structure, etc
- [ ] Comments saying why design decisions were made
- [ ] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

<!-- Describe steps to verify that the change works. -->

## Documentation changes

<!-- How it affects user workflow (CLI or API). -->

## Links

<!-- Link to all relevant specification, documentation, bug, issue or JIRA card. -->

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/

**Jira card:** JUJU-

tlm commented:
I am not sold on this. I don't understand why this is needed based on the PR description. My initial thoughts is that this shouldn't be a struct within Juju. The better way to represent this information long term is to model it as a URL that describes the source of the user. We can then have a default case of no source means the user is local. This will allow us to transition into an OIDC world with much more ease.

Aflynn50 commented:
> I am not sold on this. I don't understand why this is needed based on the PR description. My initial thoughts is that this shouldn't be a struct within Juju. The better way to represent this information long term is to model it as a URL that describes the source of the user. We can then have a default case of no source means the user is local. This will allow us to transition into an OIDC world with much more ease.

Hey, I hadn't written a proper description now, let me know what you think of the full reasoning.

In terms of a long term solution, even though this doesn't look much like OIDC, this is at least a help, in that its a lot easier to change the name and implementation of a type than it is to work with a string. Theoretically I suppose we could in future add more methods to this type such to make it work as a OIDC user ID, and it would be a lot less effort than switching from using a string everywhere.

SimonRichardson commented:
>  The better way to represent this information long term is to model it as a URL that describes the source of the user.

The consequence of that is that you need to parse out the data multiple times. Alternatively, this approach allows us to normalise into a struct. This seems like a good compromise for now. If we do want to move to OIDC this struct will nicely wrap it. Additional information can be added to the struct instead of passing additional information through the context, which I'm not a fan of having worked with at scale previously.

SimonRichardson commented on core/user/user.go:
```diff
@@ -67,3 +72,72 @@ func (u UUID) Validate() error {
+	// name is the full username.
+	name string
+	// isLocal indicates if the user is local or external.
+	isLocal bool
```
Can we store the domain on Name struct, I suspect if @tlm is right, then we'll want that information.

Aflynn50 replied:
We do have the information contained in the name already, but for consitancy it makes sense to store it in the same way as the user tag.

One thing I'm now a bit unsure about how much we should mirror the tag. The issue is that for the UserTag we have `Name()` and `Id()`. Imo this is needlessly confusing and on `user.Name` there should just be one method `Name()` that returns the full `someone@somewhere`. But this will now mean that they are inconsistent (also that `Name()` doesn't just return `n.name`). I would rather only have one method to get the "name" from the username and I would rather it was not called `Id`. What are your thoughts on this?

I could just remove the `Name` method altogether and just have only `String`. That would remove ambiguity

SimonRichardson commented on core/user/user.go:
```diff
@@ -67,3 +72,72 @@ func (u UUID) Validate() error {
+	// The empty string in an invalid user name so the struct is uninitiated if
+	// it is empty.
+	return un.name == ""
+}
```
Add a Domain method to the Name type.

Aflynn50 replied:
Added

SimonRichardson commented on core/user/user.go:
```diff
@@ -67,3 +72,72 @@ func (u UUID) Validate() error {
+func NameFromTag(tag userNameTag) Name {
+	return Name{
+		name:    tag.Id(),
+		isLocal: tag.IsLocal(),
```
names.UserTag also has a domain method, so this will mean the core/user type doesn't lose any information when moving to between the two types if we ever need to do so.

Aflynn50 replied:
We don't actually loose any information because it is all contained within the name string, but its a fair point that we might as well keep the consistant

SimonRichardson replied:
I think that's a good idea.

SimonRichardson commented on core/user/user.go:
```diff
@@ -67,3 +72,72 @@ func (u UUID) Validate() error {
+func (un Name) IsZero() bool {
+	// The empty string in an invalid user name so the struct is uninitiated if
+	// it is empty.
+	return un.name == ""
```
Maybe also check that `&& !un.isLocal`

Aflynn50 replied:
Added a check that the domain is empty

SimonRichardson commented on domain/access/state/types.go:
```diff
@@ -54,17 +55,28 @@ type dbUser struct {
 // toCoreUser converts the state user to a core user.
-func (u dbUser) toCoreUser() coreuser.User {
+func (u dbUser) toCoreUser() (coreuser.User, error) {
+	name, err := coreuser.NewName(u.Name)
+	if err != nil {
+		return coreuser.User{}, errors.Annotate(err, "user name from db")
+	}
+	var creatorName coreuser.Name
+	if u.CreatorName != "" {
+		creatorName, err = coreuser.NewName(u.CreatorName)
+		if err != nil {
+			return coreuser.User{}, errors.Annotate(err, "creator name from db")
+		}
+	}
```
If this does error out here, I wonder what the path forward to fixing it would be. I guess direct DB access to rectify it.

Aflynn50 replied:
Yea, as @tlm has said before, we should have the guarantee that the stuff in the DB is valid and good when we add it. I would kindof like not to have to do validation at all here, but that would mean having a seperate method to create a username without checking if its valid, and I don't really want to do that. This seems like an ok compromise

SimonRichardson commented on domain/cloud/state/state.go:
```diff
@@ -898,8 +899,8 @@ ON CONFLICT(cloud_uuid, name) DO UPDATE SET name=excluded.name,
-func insertPermission(ctx context.Context, tx *sqlair.TX, ownerName, cloudName string) error {
-	if ownerName == "" {
+func insertPermission(ctx context.Context, tx *sqlair.TX, ownerName user.Name, cloudName string) error {
+	if ownerName.Name() == "" {
```
Surely this is `ownerName.IsZero()`?

Aflynn50 replied:
Good point, have done a grep and fixed all the other places this slipped though.

SimonRichardson commented on domain/model/modelmigration/import.go:
```diff
@@ -159,7 +159,7 @@ func (i *importOperation) Execute(ctx context.Context, model description.Model)
 		cred.Name = model.CloudCredential().Name()
 		cred.Cloud = model.CloudCredential().Cloud()
-		cred.Owner = model.CloudCredential().Owner()
+		cred.Owner, _ = coreuser.NewName(model.CloudCredential().Owner())
```
We shouldn't ignore this error.

Aflynn50 replied:
Looks like that slipped through with the IDE automatic refactoring, have grepped for other instances and it looks like that was the only one.

SimonRichardson commented on domain/model/state/modelstate.go:
```diff
@@ -147,6 +149,13 @@ FROM model
 		return coremodel.ReadOnlyModel{}, errors.Trace(err)
 	}
 
+	if credOwnerName != "" {
```
It might be worth adding an info log to state that the owner is empty.

Aflynn50 replied:
Added, had to pipe a logger in

SimonRichardson reviewed (commented):
Thanks for the changes, I like the way this is going. Although @tlm has concerns about just using a string URL for OIDC, I think a normalized struct with additional data is preferable from my perspective.

tlm commented:
I still have a number of concerns about this PR that isn't well thought out for the use cases we have. I haven't done a great job at explaining myself.

If we look at Juju today we have 4 sources of users. That is users can come from:
- local controller
- remote controller
- external macaroon sources
- JAAS

In the future it is without out a doubt that we also possible get users from:
- OIDC
- External PKI (x509)
- Spiffe, closely related to the above.

To represent all of these different users and how we got them we should adopt a URI or URL pattern for them.

Some examples of this would be:
- juju://tlm@<controller-id>
- juju://tlm@external-controller-ip
- macaroon://tlm@external-source
- jaas://wallyworld@jaas-wll-known-path
- ....

We can determine if a user is a local user by the schema and host portion. All thought I am not sure we want code through out Juju that has `if isLocal` logic.

My main concern is that we are not modelling the problem correctly and by having specific bools to tell us properties of a users source is missing the bar long term.

Aflynn50 commented:
I think understand the points being made here. @tlm I agree that this is definatly not the ideal solution, and switching on the user with `.IsLocal` is never really what we want to do. As @SimonRichardson suggested, it makes sense that we would just pass in different services for this. I have already added in previous PRs a lot of horrible code to the access state that effectively special cases external users and hopefully can be removed when we address this properly. 

My main motivation for this PR was [this line here](https://github.com/juju/juju/pull/17775/files#diff-3e23bf35fdd67fdf04daf1262e1c73e5cd3e8bd8394fddc8f153cd0052dcd184L333) where we need to check that if the user is external in the domain. I really didn't like using the names package.

Imo it will still be very useful to have this username type even if we remove the `IsLocal` method now. The solution @SimonRichardson suggested addresses the issues of authn/authz, this would be passed out to the service passed into the api facade. But users don't only exists there in juju. If an external user is registered as a user on a model it needs to be listed when you ask for the model users. I'm sure there are other examples as well, but I believe that juju controller needs to have some record of external users even if it is not doing authn/authz for them. This type can be a placeholder for that.

Have a type for this instead of a string also makes future refactoring a ton easier as well. I've ended up putting several full days of work into this and so if it gets merged we can always use IDE refactoring to change the method names on it, or change how we store the username. This PR was done with the assumption that we will overhaul our permissions system and in the knowledge that the implementation is not currently ideal. 

I can also put some comments/todos around the IsLocal method explaining that it is not something that should be used unless absolutely necessary. I don't want to add the temptation to use conditional logic to handle this stuff everywhere.

SimonRichardson reviewed (approved):
I think this is a good step forward. It's not going to be our ideal solution. That requires a change to how we would do authz/authn (untangling them). So for now, send it and understand that we will have to revisit this again.

Aflynn50 commented:
/merge

manadart commented on state/modeluser_test.go:
```diff
@@ -45,7 +46,7 @@ func (s *ModelUserSuite) TestAddModelUser(c *gc.C) {
 	c.Assert(modelUser.UserID, gc.Equals, fmt.Sprintf("%s:validusername", s.modelTag.Id()))
 	c.Assert(modelUser.Object, gc.Equals, s.modelTag)
-	c.Assert(modelUser.UserName, gc.Equals, "validusername")
+	c.Assert(modelUser.UserName, gc.Equals, usertesting.GenNewName(c, "validusername"))
```
These are all superfluous changes.
- Generating valid names is useful for input, but that should be validated in logic being tested. There's no need to run it through for result comparison.
- These packages are all disappearing.

manadart commented on state/useraccess.go:
```diff
@@ -152,7 +153,11 @@ func (st *State) UserPermission(subject names.UserTag, target names.Tag) (permis
 }
 
-func newUserAccess(perm *userPermission, userDoc userAccessDoc, object names.Tag) permission.UserAccess {
+func newUserAccess(perm *userPermission, userDoc userAccessDoc, object names.Tag) (permission.UserAccess, error) {
```
None of this is worth changing in the state package.
